### DISCUSSION TITLE
Table custom filter - not working well with Select

### DISCUSSION THREAD

**Discussion by HaveSpacesuit** [iTwin/iTwinUI-react#658] [Q&A] [answered]
I'm trying to implement a custom filter in the table using a Select dropdown. Every time the user tries to click an option, the filter closes, and the filter is not applied. Other elements such as radio boxes work, but a Select or ComboBox would be preferable, as we will have long lists.

Relatedly, for a boolean filter, if filtering the table for `false` values, the filter icon is not highlighted. Not sure if this is a React-Table bug, an iTwinUI-react bug, or some additional tweaking I need to take care of.

See this code sandbox. The "supported" column uses a Select, and does not work. The "deprecated" column uses "OptionGroup" and works, though the filter is not highlighted when filtering for false.

https://codesandbox.io/s/itwinui-react-example-forked-ww4opp

**Comment by mayank99** [comment, marked as answer]
Hi @HaveSpacesuit, thanks for creating this discussion.

The problem here is that our popovers are configured to automatically close when clicked "outside". In this case, the dropdown from `Select` is placed in a portal as a child of `<body>`, so it is considered to be "outside" the filter popover.

As a workaround, you can use `appendTo: 'parent'` to place the select dropdown in the same place as the select button. This is also somewhat documented in our [FAQ](https://github.com/iTwin/iTwinUI-react/wiki/FAQ#popover-tooltips-dropdowns-etc).
![<Select options={options} popoverProps={{ appendTo: 'parent' }} />](https://user-images.githubusercontent.com/9084735/166704252-c01acedf-d77e-4915-8cf1-9cd0c7364380.png)

Hope that helps.

**Reply by HaveSpacesuit** [reply]
Wonderful, thanks!

It looks like we can improve the filter state by changing the filter logic to specifically exclude undefined:

```diff
// Table.tsx #510
-const areFiltersSet = allColumns.some((column) => !!column.filterValue);
+const areFiltersSet = allColumns.some((column) => column.filterValue !== undefined);

// FilterToggle.tsx
-<IconButton isActive={isVisible || column.filterValue}>
-  {column.filterValue ? <SvgFilter /> : <SvgFilterHollow />}
-</IconButton>
+<IconButton isActive={isVisible || column.filterValue !== undefined}>
+  {column.filterValue !== undefined ? <SvgFilter /> : <SvgFilterHollow />}
+</IconButton>

```

**Reply by mayank99** [reply]
I agree, it should include `false`. I would also exclude `null` by using `!=` instead of `!==`.

Do you want to create a PR for this?

**Reply by HaveSpacesuit** [reply]
Sure, I'd be happy to.

**Reply by veekeys** [reply]
This is released in [1.38.0](https://github.com/iTwin/iTwinUI-react/releases/tag/v1.38.0)
